src/vm/marshalnative.cpp

```diff
@@ -923,6 +923,133 @@ FCIMPL1(int, MarshalNative::GetHRForException_WinRT, Object* eUNSAFE)
+
+    BEGIN_QCALL;
+
+    handle = NDirect::LoadLibraryFromPath(path, throwOnError);
+
+    END_QCALL;
+
+    return reinterpret_cast<INT_PTR>(handle);
+}
+
+// static
+INT_PTR QCALLTYPE MarshalNative::LoadLibraryByName(LPCWSTR name, QCall::AssemblyHandle callingAssembly, 
+                                                   DWORD dllImportSearchPathFlag, BOOL throwOnError)
+{
+    QCALL_CONTRACT;
+
+    NATIVE_LIBRARY_HANDLE handle = nullptr;
+    Assembly *pAssembly = callingAssembly->GetAssembly();
+
+    BEGIN_QCALL;
+
+    handle = NDirect::LoadLibraryByName(name, pAssembly, dllImportSearchPathFlag, throwOnError);
+
+    END_QCALL;
+
+    return reinterpret_cast<INT_PTR>(handle);
+}
+
+// static
+void QCALLTYPE MarshalNative::FreeNativeLibrary(INT_PTR handle, BOOL throwOnError)
+{
+    QCALL_CONTRACT;
+    BOOL retVal = 0;
+
+    BEGIN_QCALL;
+
+    if (handle == NULL)
+    {
+        if(throwOnError)
+            COMPlusThrow(kArgumentNullException, W("ArgumentNull_handle"));
+        goto lExit;
+    }
+
+#ifndef FEATURE_PAL
+    retVal = FreeLibrary((HMODULE)handle);
+#else // !FEATURE_PAL
+    retVal = PAL_FreeLibraryDirect((NATIVE_LIBRARY_HANDLE)handle);
+#endif // !FEATURE_PAL
+
+    if (throwOnError && (retVal == 0))
+    {
+        COMPlusThrow(kInvalidOperationException, W("ArgumentNull_handle"));
+    }
+    
+lExit: ;
+    END_QCALL;
+}
+
+//static 
+INT_PTR QCALLTYPE MarshalNative::GetNativeLibraryExport(INT_PTR handle, LPCWSTR symbolName, BOOL throwOnError)
+{
+    QCALL_CONTRACT;
+
+    INT_PTR address = NULL;
+#ifndef FEATURE_PAL
+    NewArrayHolder<char> lpstr = NULL;
+#endif // !FEATURE_PAL
+
+    BEGIN_QCALL;
+    
+    if (handle == NULL)
+    {
+        if(throwOnError)
+            COMPlusThrow(kArgumentNullException, W("ArgumentNull_handle"));
+        goto lExit;
+    }
+
+    if (symbolName == NULL)
+    {
+        if(throwOnError)
+            COMPlusThrow(kArgumentNullException, W("ArgumentNull_symbolName"));
+        goto lExit;
+    }
+
+#ifndef FEATURE_PAL
+    // GetProcAddress only takes  LPCSTR for procedure name.
+    // Therefore, convert symbolName to a multi-byte-char encoding before calling GetProcAddress()
+    DWORD wideLen = (DWORD) wcslen(symbolName);
+    DWORD maxMbLen = (wideLen + 1) * 3;
+    lpstr = new char[maxMbLen];
+    DWORD mbLen = WszWideCharToMultiByte(CP_UTF8, 0, symbolName, -1, lpstr, maxMbLen, nullptr, nullptr);
+
+    if (mbLen == 0)
+    {
+        if(throwOnError)
+            COMPlusThrow(kEntryPointNotFoundException, IDS_EE_NDIRECT_GETPROCADDRESS_WIN, W("GetLibraryExport_handle"), symbolName);
+        goto lExit;
+    }
+
+    address = reinterpret_cast<INT_PTR>(GetProcAddress((HMODULE)handle, lpstr));
+#else // !FEATURE_PAL
+    // PAL_GetProcAddressDirect() accepts wide-string procedure name, 
+    // in line with other public PAL API.
+
+    address = reinterpret_cast<INT_PTR>(PAL_GetProcAddressDirect((NATIVE_LIBRARY_HANDLE)handle, symbolName));
```

Comment:
It would be better to `PAL_GetProcAddressDirect` to also take the entrypoint as LPCSTR. It would reduce amount of ifdefed out code.